core/domain/learner_playlist_services.py

```diff
@@ -0,0 +1,240 @@
+
+from core.domain import subscription_services
+from core.domain import user_domain
+from core.platform import models
+
+import feconf
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+MAX_LEARNER_PLAYLIST_ACTIVITY_COUNT = (
+    feconf.MAX_LEARNER_PLAYLIST_ACTIVITY_COUNT)
+
+
+def get_learner_playlist_from_model(learner_playlist_model):
+    """Returns the learner playlist domain object given the learner playlist
+    model loaded from the datastore.
+
+    Args:
+        learner_playlist_model: LearnerPlaylistModel. The
+            learner playlist model from the datastore.
+
+    Returns:
+        LearnerPlaylist. The learner playlist domain object corresponding to the
+            given model.
+    """
+    return user_domain.LearnerPlaylist(
+        learner_playlist_model.id,
+        learner_playlist_model.exploration_ids,
+        learner_playlist_model.collection_ids)
+
+
+def save_learner_playlist(learner_playlist):
+    """Save a learner playlist domain object as an LearnerPlaylistModel entity
+    in the datastore.
+
+    Args:
+        learner_playlist: LearnerPlaylist. The learner playlist domain object to
+            be saved in the datastore.
+    """
+    learner_playlist_model = user_models.LearnerPlaylistModel(
+        id=learner_playlist.id,
+        exploration_ids=learner_playlist.exploration_ids,
+        collection_ids=learner_playlist.collection_ids)
+
+    learner_playlist_model.put()
+
+
+def mark_exploration_to_be_played_later(
+        user_id, exploration_id, position_to_be_inserted=None):
+    """Adds the exploration id to the learner playlist of the user at the given
+    position. If the position is not specified, the exploration gets added at
+    the end. If the exploration is one of the subscribed explorations of the
+    user, it is not added. The maximum limit of the learner playlist is
+    feconf.MAX_LEARNER_PLAYLIST_ACTIVITY_COUNT. If the count exceeds
+    feconf.MAX_LEARNER_PLAYLIST_ACTIVITY_COUNT, the exploration is not added.
+
+    Args:
+        user_id: str. The id of the user.
+        exploration_id: str. The id of the exploration to be added to the
+            learner playlist.
+        position_to_be_inserted: int|None. If this is specified the exploration
+            gets inserted at the given position. Otherwise it gets added at the
+            end.
+    """
+    learner_playlist_model = user_models.LearnerPlaylistModel.get(
+        user_id, strict=False)
+    if not learner_playlist_model:
+        learner_playlist_model = (
+            user_models.LearnerPlaylistModel(id=user_id))
+
+    subscribed_exploration_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    learner_playlist = get_learner_playlist_from_model(
+        learner_playlist_model)
+
+    if exploration_id not in subscribed_exploration_ids:
+        exploration_ids_count = len(learner_playlist.exploration_ids)
+        if not position_to_be_inserted:
```

Comment:
I don't think the approach would work. In many cases, we would have to call this function where we wouldn't have access to the entire list. For example when the user clicks the clock icon on the exploration in the library - only that exploration should be added. 

Secondly, this is the more general function - it can be used in the case of rearranging the activities as well as when the user wants to add an individual activity. So this serves both the purposes.

Thirdly, even though payload may not be a problem this function is still more efficient and it might make a difference in places with limited bandwidth. Also, the number 10 has been kept after just a few initial user interviews. We might have to increase the number later. Maybe even remove the limit.

Regarding your point about this function being error prone we can always make sure that this function is error free as this is just one small function. Also, I have added more tests to make sure nothing goes wrong. 

Does this explanation suffice?